feat: add support for SummaryDataPoint in delta to cumulative processor

lightme16 opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
Adding support for Delta to Cumulatives for summary data points 
<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38772

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Added new unit tests for summary

<!--Describe the documentation added.-->
#### Documentation

<!--Please delete paragraphs that you did not use before submitting.-->

atoulme commented:
Please add tests to cover the new functionality?

lightme16 commented:
> Please add tests to cover the new functionality?

@atoulme, I can definitely do that. Thank you! However, before I invest further, could you please review the business logic? Are you satisfied with this functionality? Do you agree that sum & count metrics in the summary should be aggregated? I believe they should be, especially for our use case. However, I’m not sure if there are known issues that prevented this from being implemented in the past. If this functionality is indeed missing, we’re happy to add it, but I want to confirm with the community first.

More in the https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38772

lightme16 committed:
feat: add support for SummaryDataPoint in delta to cumulative processor

lightme16 commented on processor/deltatocumulativeprocessor/processor.go:
```diff
@@ -83,10 +85,11 @@ func (p *Processor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) erro
 
 	metrics.Filter(md, func(m metrics.Metric) bool {
-		if m.AggregationTemporality() != pmetric.AggregationTemporalityDelta {
+		if m.AggregationTemporality() != pmetric.AggregationTemporalityDelta && m.Type() != pmetric.MetricTypeSummary {
```
I am uncertain about this part. The summary has `AggregationTemporalityUnspecified`, which is why this check skipped it. However, I believe we need to address it, so I added this condition.

lightme16 commented:
> Please add tests to cover the new functionality?

Added a test

atoulme commented on processor/deltatocumulativeprocessor/processor.go:
```diff
@@ -161,6 +164,27 @@ func (p *Processor) ConsumeMetrics(ctx context.Context, md pmetric.Metrics) erro
 					err = p.aggr.Exponential(last, dp)
 					last.CopyTo(dp)
 				})
+			// Add a case for SummaryDataPoint in the switch statement
```
```suggestion
```

lightme16 committed:
feat: add summary test cases and update existing test data for delta to cumulative processor

lightme16 commented on processor/deltatocumulativeprocessor/testdata/notemporality-ignored/1.test:
```diff
@@ -25,13 +25,6 @@ resourceMetrics:
               dataPoints:
                 - timeUnixNano: 1
                   asDouble: 1
-          - name: test.summary
```
I don't think this test is relevant anymore as we started to process summary data points. 
I added new testdata/summary dir to unit test the new logic

sh0rez commented on processor/deltatocumulativeprocessor/internal/data/summ_test.go:
```diff
@@ -0,0 +1,197 @@
+}
+
+// equalSummaryDataPoints compares two summary data points, treating nil and empty quantile slices as equivalent
+func equalSummaryDataPoints(expected, actual pmetric.SummaryDataPoint) bool {
```
why is this custom logic required? what makes `datatest.Is.Equal` insufficient?

lightme16 replied:
I just removed that function. The idea was to compare quantiles only when they exist, otherwise they would be null.

tombrk reviewed (changes requested):
hey @lightme16, thanks for your contribution!

after reading https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38772, I can see where you are coming from.

however, reading the [metrics spec](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#summary-legacy), it seems OTel summaries have no temporality attached .. even though your datadog conversion appears to put such data into the data structure.

if we were to aggregate .sum and .count of summaries, we would be operating outside the spec afaict, which I am not sure we should / could.
if we were to decide in favor of that, I think this would need to be at least behind a default-off config option.

opinions? @ArthurSens @tedsuo

lightme16 commented:
Hey, @sh0rez, thank you for your review. I agree that it could be unclear, but from a practical standpoint, it seems to be an essential feature when using StatsD receiver and Prometheus exporters.

>  even though your datadog conversion appears to put such data into the data structure.

I want to clarify that it's not our DataDog conversion that adds those sum and count delta counters; it's simply how OpenTelemetry agents operate. When using the Prometheus exporter, it converts OTLP summary data structures into sum count data structures and another gauge with percentiles. Eg:

```
exporters:
    prometheus:
        endpoint: "0.0.0.0:8800"
    deltatocumulative:

receivers:
    statsd:
        aggregation_interval: 15s
        endpoint: 0.0.0.0:8135
        timer_histogram_mapping:
            - observer_type: summary
              statsd_type: distribution
              summary:
                  percentiles:
                      - 50
                      - 99
service:
    pipelines:
        metrics/statsd:
            receivers:
                - statsd
            processors:
                - deltatocumulative
            exporters:
                - prometheus 
````

I don't see how anyone could utilize that when using the StatsD receiver and Prometheus exporter because Prometheus doesn't support delta counters. This is what OTLP summary data structures are converted to, so I would argue that if you use the StatsD receiver and Prometheus exporter, those count and sum metrics derived from summary data points are useless as they are reseted every `aggregation_interval` eg 15s. I'm not sure how one could rely on this data.


The problem we face, in simple terms, is that we have DataDog distributions (e.g., HTTP request latency):

`echo -n "http.requestTime.dist:100|d|#region:us-west-2" > /dev/udp/localhost/8125`

From that metric, we can obtain latency values like median, P99, etc. We also use it to calculate RPS (requests per second) if we use `_count` metric:
`sum(rate(http_requestTime_count{service=~"foo"}[$__rate_interval])) by (region)`

However, if you check the screenshot you can see that on (1) datapoints is spiky as metrics are reset every 15s. 


Just today I created a custom OpenTelemetry binary from my merge request and successfully tested it in our company environment, and it has been working well so far. Please see (2) on a screenhost. Prometheus RPS queries are in sync with the Datadog queries we are using. Please see screenhot:
![CleanShot 2025-03-20 at 22 43 05@2x](https://github.com/user-attachments/assets/cb4db8e9-f475-49bb-b403-1281d18506c4)


Therefore, I believe this feature is absolutely necessary, and it should be enabled by default. However, it's up to you all to decide if you think an optional flag is needed. I can easily add it; just let me know if I should proceed.

Thank you!

lightme16 committed:
feat: add accumulate_summaries configuration to delta to cumulative processor

feat: add changelog entry for cumulative sum support in delta to cumulative processor

lightme16 commented:
@sh0rez, @ArthurSens, and @tedsuo: I just added a feature flag to disable the summary accumulation functionality by default. Please share your thoughts on the merge request. My company needs this to move forward, and it would be great to have it merged. Let me know if anything else broke while processing this for merging.

Thanks!

lightme16 committed:
Merge branch 'main' into deltatocumulativeprocessor-summary

lightme16 committed:
Merge branch 'main' into deltatocumulativeprocessor-summary

lightme16 commented:
hello, any thoughts on MR - @sh0rez, @ArthurSens, and @tedsuo? thanks

sh0rez commented:
hey @lightme16, I apologize for the much too long waiting time. I was on pto and did not get back to this earlier.

while I _do_ think you are onto a very valid use-case we should absolutely enable, I am not comfortable with deltatocumulative relying on behavior that is not part of the specification. 

as per the spec, the [Summary data point](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#summary-legacy) has no aggregation temporality. your example however shows it absolutely should have such information attached, because `.sum` and `.count` can clearly be delta.

ideally, I imagine the following:
- spec change to make `SummaryDataPoint` have `AggregationTemporality`
- make `statsdreceiver` set above to delta
- adapt and merge this PR

are you interested in continuing this work?

lightme16 commented:
Hello, @sh0rez  I decided to move away from that summaries approach in my system due to this issue and use new native Prometheus histograms instead. I'm not interested in doing any more work on this, but I still believe it would be beneficial to have this patch merged and implemented. Perhaps someone else could take it on
